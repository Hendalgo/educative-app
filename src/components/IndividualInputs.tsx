import React from 'react';
import {View, TextInput, StyleProp} from 'react-native';
import {useTheme} from '@react-navigation/native';
import styles from '@styles/index';
import {CustomColors} from '@styles/themes';

interface IndividualInputsProps {
  values: number[];
  setValues: (values: number[]) => void;
  elementsNum: number;
  onCompleted?: (completed: boolean) => void;
}

const IndividualInputs = ({
  values,
  setValues,
  elementsNum,
  onCompleted,
}: IndividualInputsProps): React.JSX.Element => {
  //@ts-ignore
  const {colors}: {colors: CustomColors} = useTheme();
  const inputRefs = React.useRef<any>([]);
  const [stylesInputs, setStylesInputs] = React.useState<Array<StyleProp<any>>>(
    Array.from({length: elementsNum}, () => ({
      backgroundColor: colors.neutral000,
      borderColor: colors.neutral300,
      color: colors.neutral600,
    })),
  );
  const handleChanges = (text: string, index: number) => {
    const newValues: any = [...values];
    if (text) {
      newValues[index] = parseInt(text, 10);
    } else {
      newValues.splice(index, 1);
    }
    setValues(newValues);
    if (text.length > 0) {
      setStylesInputs([
        ...stylesInputs.slice(0, index),
        {
          ...stylesInputs[index],
          borderColor: colors.primary000,
          color: colors.black,
        },
        ...stylesInputs.slice(index + 1),
      ]);
    } else {
      setStylesInputs([
        ...stylesInputs.slice(0, index),
        {
          ...stylesInputs[index],
          borderColor: colors.neutral300,
          color: colors.black,
        },
        ...stylesInputs.slice(index + 1),
      ]);
    }
    if (newValues.length === elementsNum && onCompleted) {
      onCompleted(true);
    } else {
      if (onCompleted) {
        onCompleted(false);
      }
    }
  };
  const handleFocus = (nativeEvent: any, index: number) => {
    //Move focus to the next or previous empty input
    //If it is, focus on it
    const indexTofocus =
      index === 0 ? index : index === values.length - 1 ? index : values.length;
    inputRefs.current[indexTofocus].focus();
  };
  const handleKeyPress = (event: any, index: number) => {
    if (event.nativeEvent.key === 'Backspace' && values[index] === undefined) {
      if (index > 0) {
        // Cambiar el foco al input anterior
        inputRefs.current[index - 1].focus();

        // Eliminar el valor del input anterior
        const newValues = [...values];
        newValues.splice(index - 1, 1);
        setValues(newValues);

        // Establecer el estilo del input anterior como vacÃ­o
        setStylesInputs(prevStyles => [
          ...prevStyles.slice(0, index - 1),
          {
            ...prevStyles[index - 1],
            borderColor: colors.neutral300,
            color: colors.black,
          },
          ...prevStyles.slice(index),
        ]);
      }
    }
    //if fill the last input, move to the next input
    if (event.nativeEvent.key !== 'Backspace' && index < elementsNum - 1) {
      inputRefs.current[index + 1].focus();
    }
  };
  return (
    <View
      style={[
        styles.individualInputsContainer,
        {
          backgroundColor: colors.neutral000,
        },
      ]}>
      {Array.from({length: elementsNum}, (_, i) => {
        return (
          <TextInput
            testID={`individualInput-${i}`}
            key={i}
            style={[
              styles.individualInput,
              {
                borderColor: stylesInputs[i]?.borderColor
                  ? stylesInputs[i].borderColor
                  : colors.neutral300,
                backgroundColor: stylesInputs[i]?.backgroundColor
                  ? stylesInputs[i].backgroundColor
                  : colors.card,
                color: stylesInputs[i]?.color
                  ? stylesInputs[i].color
                  : colors.neutral1500,
              },
            ]}
            placeholder={`${i + 1}`}
            placeholderTextColor={colors.neutral1500}
            keyboardType="numeric"
            maxLength={1}
            value={values[i] ? values[i].toString() : ''}
            onChangeText={(text: string) => handleChanges(text, i)}
            ref={ref => (inputRefs.current[i] = ref)}
            onFocus={({nativeEvent}) => handleFocus(nativeEvent, i)}
            onKeyPress={event => handleKeyPress(event, i)}
          />
        );
      })}
    </View>
  );
};

export default IndividualInputs;
