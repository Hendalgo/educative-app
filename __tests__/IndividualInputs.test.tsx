import React from 'react';
import IndividualInputs from '@components/IndividualInputs';
import {fireEvent, render} from '@testing-library/react-native';
import {useState} from 'react';

const IndividualInputsTest = ({elementsNum = 4}: {elementsNum?: number}) => {
  const [values, setValues] = useState<number[]>([]);
  return (
    <IndividualInputs
      elementsNum={elementsNum}
      values={values}
      setValues={setValues}
    />
  );
};

describe('IndividualInputs', () => {
  it('renders correctly', () => {
    const tree = render(<IndividualInputsTest />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly with 6 elements', () => {
    const tree = render(<IndividualInputsTest elementsNum={6} />);
    const inputs = tree.getAllByTestId(/individualInput-/);
    expect(inputs.length).toBe(6);
  });
  it('changes the values correctly', () => {
    const {getByTestId} = render(<IndividualInputsTest />);
    const input = getByTestId('individualInput-0');
    fireEvent.changeText(input, '1');
    expect(input.props.value).toBe('1');
  });
  it('completed the inputs correctly', () => {
    const {getByTestId} = render(<IndividualInputsTest />);
    const input = getByTestId('individualInput-0');
    fireEvent.changeText(input, '1');
    const input2 = getByTestId('individualInput-1');
    fireEvent.changeText(input2, '2');
    const input3 = getByTestId('individualInput-2');
    fireEvent.changeText(input3, '3');
    const input4 = getByTestId('individualInput-3');
    fireEvent.changeText(input4, '4');
    expect(input4.props.value).toBe('4');
  });
});
