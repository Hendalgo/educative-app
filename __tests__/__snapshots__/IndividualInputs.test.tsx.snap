// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndividualInputs renders correctly 1`] = `
<View
  style={
    [
      {
        "alignContent": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
      {
        "backgroundColor": undefined,
      },
    ]
  }
>
  <TextInput
    keyboardType="numeric"
    maxLength={1}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="1"
    style={
      [
        {
          "borderRadius": 8,
          "borderWidth": 1,
          "fontFamily": "Poppins-Regular",
          "fontSize": 16,
          "marginHorizontal": 5,
          "marginVertical": 10,
          "padding": 12,
          "textAlign": "center",
          "width": 50,
        },
        {
          "backgroundColor": "rgb(255, 255, 255)",
          "borderColor": undefined,
          "color": undefined,
        },
      ]
    }
    testID="individualInput-0"
    value=""
  />
  <TextInput
    keyboardType="numeric"
    maxLength={1}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="2"
    style={
      [
        {
          "borderRadius": 8,
          "borderWidth": 1,
          "fontFamily": "Poppins-Regular",
          "fontSize": 16,
          "marginHorizontal": 5,
          "marginVertical": 10,
          "padding": 12,
          "textAlign": "center",
          "width": 50,
        },
        {
          "backgroundColor": "rgb(255, 255, 255)",
          "borderColor": undefined,
          "color": undefined,
        },
      ]
    }
    testID="individualInput-1"
    value=""
  />
  <TextInput
    keyboardType="numeric"
    maxLength={1}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="3"
    style={
      [
        {
          "borderRadius": 8,
          "borderWidth": 1,
          "fontFamily": "Poppins-Regular",
          "fontSize": 16,
          "marginHorizontal": 5,
          "marginVertical": 10,
          "padding": 12,
          "textAlign": "center",
          "width": 50,
        },
        {
          "backgroundColor": "rgb(255, 255, 255)",
          "borderColor": undefined,
          "color": undefined,
        },
      ]
    }
    testID="individualInput-2"
    value=""
  />
  <TextInput
    keyboardType="numeric"
    maxLength={1}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    placeholder="4"
    style={
      [
        {
          "borderRadius": 8,
          "borderWidth": 1,
          "fontFamily": "Poppins-Regular",
          "fontSize": 16,
          "marginHorizontal": 5,
          "marginVertical": 10,
          "padding": 12,
          "textAlign": "center",
          "width": 50,
        },
        {
          "backgroundColor": "rgb(255, 255, 255)",
          "borderColor": undefined,
          "color": undefined,
        },
      ]
    }
    testID="individualInput-3"
    value=""
  />
</View>
`;
